File: modules/Scripting/annotations/RmlUi/Tween.lua
---@meta

---@class ui.Tween
---@overload fun(type?: ui.Tween.Type, direction?: ui.Tween.Direction): ui.Tween
---@overload fun(typeIn: ui.Tween.Type, typeOut: ui.Tween.Type): ui.Tween
ui.Tween = {}

---@enum ui.Tween.Type
ui.Tween.Type = {
  None = 0,
  Back = 1,
  Bounce = 2,
  Circular = 3,
  Cubic = 4,
  Elastic = 5,
  Exponential = 6,
  Linear = 7,
  Quadratic = 8,
  Quartic = 9,
  Quintic = 10,
  Sine = 11,
  Callback = 12,
  Count = 13,
}

---@enum ui.Tween.Direction
ui.Tween.Direction = {
  In = 1, Out = 2, InOut = 3,
}

---Reverse direction of the tweening function.
function ui.Tween:reverse() end

File: modules/Scripting/modules/RmlUi/src/Tween.cpp
#include "Tween.hpp"
#include "RmlUi/Core/Tween.h"

#include "Sol2HelperMacros.hpp"

using namespace Rml;

void registerTween(sol::table &lua) {
  // clang-format off
  lua.DEFINE_USERTYPE(Tween,
    sol::call_constructor,
    sol::constructors<
      Tween(),
      Tween(Tween::Type),
      Tween(Tween::Type, Tween::Direction),
      Tween(Tween::Type, Tween::Type)
    >(),

    _BIND(Tween, reverse),

    sol::meta_function::to_string, &Tween::to_string
  );

#define MAKE_PAIR(Value) _MAKE_PAIR(Tween::Type, Value)
  lua DEFINE_NESTED_ENUM(Tween, Type, {
    MAKE_PAIR(None),
    MAKE_PAIR(Back),
    MAKE_PAIR(Bounce),
    MAKE_PAIR(Circular),
    MAKE_PAIR(Cubic),
    MAKE_PAIR(Elastic),
    MAKE_PAIR(Exponential),
    MAKE_PAIR(Linear),
    MAKE_PAIR(Quadratic),
    MAKE_PAIR(Quartic),
    MAKE_PAIR(Quintic),
    MAKE_PAIR(Sine),
    MAKE_PAIR(Callback),
    MAKE_PAIR(Count),
  });
#undef MAKE_PAIR

#define MAKE_PAIR(Value) _MAKE_PAIR(Tween::Direction, Value)
  lua DEFINE_NESTED_ENUM(Tween, Direction, {
    MAKE_PAIR(In),
    MAKE_PAIR(Out),
    MAKE_PAIR(InOut),
  });
#undef MAKE_PAIR
  // clang-format on
}
